### PR TITLE
Improve algorithm to detect dependency changes

### PR DESCRIPTION
**Pull request by reuzel**
Hi,

Found a couple of cases where dependencies were not updated correctly. Especially in corner-cases with cyclic dependencies. Added one of them as test-case.

This pull requests replaces the (depth-first) recursive search for changes with a repeated breadth-first approach. How it works: a loop is inserted that triggers release type updates of all packages. In each cycle each package will check its own dependencies (only) and updates its release type if necessary. With each cycle changes are propagated "upward" the dependency graph. The process stops when no package is changing its release type anymore: the graph is stable. This process has no "ignore" lists as part of the search, and as such is better able to capture changes in complex dependency graphs.

Hope this helps,
Joost

### CODE REVIEW

File: lib/updateDeps.js [hunk 1]
```diff
@@ -130,52 +130,61 @@ const _nextPreHighestVersion = (latestTag, lastVersion, pkgPreRelease) => {
...
+		pkg._depsChanged.every((dep) => dep._lastRelease && !dep._nextType) //no new deps or deps upgraded
+	)
+		return pkg._nextType;
+
+	//find highest release type if strategy is inherit, starting of type set by commit analyzer
+	if (releaseStrategy === "inherit") {
+		return pkg._depsChanged.reduce((maxReleaseType, dependency) => {
+			return severityOrder.indexOf(dependency._nextType) > severityOrder.indexOf(maxReleaseType)
+				? dependency._nextType
+				: maxReleaseType;
```
Comment (antongolub) [inline lib/updateDeps.js:160]: Let's introduce a helper for this:

```js
const getHighestReleaseType = (...releaseTypes) =>
    ["major", "minor", "patch"].find(type => releaseTypes.includes(type))

// ...
if (releaseStrategy === "inherit") {
   return getHighestRelease(pkg._depsChanged.map(d => d._nextType))
}

//...
return getHighestRelease(pkg._nextType, releaseStrategy)
```

Comment (reuzel) [reply lib/updateDeps.js]: done

**Comment by reuzel** [comment]
Please wait before merging. Seems there is a possibility that the loop doesn't end... checking...

**Comment by reuzel** [comment]
never mind, block had nothing to do with the changes, but with my Git setup (a token expired ;-) )

**Review by antongolub — APPROVED** [review]
